src/Security/MemberAuthenticator/CookieAuthenticationHandler.php

```diff
@@ -0,0 +1,245 @@
+    {
+        return $this->tokenCookieName;
+    }
+
+    /**
+     * Set the name of the cookie used to store an login token
+     *
+     * @param $tokenCookieName
+     * @return null
+     */
+    public function setTokenCookieName($tokenCookieName)
+    {
+        $this->tokenCookieName = $tokenCookieName;
+    }
+
+    /**
+     * Once a member is found by authenticateRequest() pass it to this identity store
+     *
+     * @return IdentityStore
+     */
+    public function getCascadeLogInTo()
+    {
+        return $this->cascadeLogInTo;
+    }
+
+    /**
+     * Set the name of the cookie used to store an login token
+     *
+     * @param $cascadeLogInTo
+     * @return null
+     */
+    public function setCascadeLogInTo(IdentityStore $cascadeLogInTo)
+    {
+        $this->cascadeLogInTo = $cascadeLogInTo;
+    }
+
+    /**
+     * @param HTTPRequest $request
+     * @return null|Member
+     */
+    public function authenticateRequest(HTTPRequest $request)
+    {
+        $uidAndToken = Cookie::get($this->getTokenCookieName());
+        $deviceID = Cookie::get($this->getDeviceCookieName());
+
+        // @todo Consider better placement of database_is_ready test
+        if (!$deviceID || strpos($uidAndToken, ':') === false || !Security::database_is_ready()) {
+            return null;
+        }
+
+        list($uid, $token) = explode(':', $uidAndToken, 2);
+
+        if (!$uid || !$token) {
+            return null;
+        }
+
+        /** @var Member $member */
+        $member = Member::get()->byID($uid);
+
+        /** @var RememberLoginHash $rememberLoginHash */
+        $rememberLoginHash = null;
+
+        // check if autologin token matches
+        if ($member) {
+            $hash = $member->encryptWithUserSettings($token);
+            $rememberLoginHash = RememberLoginHash::get()
+                ->filter(array(
+                    'MemberID' => $member->ID,
+                    'DeviceID' => $deviceID,
+                    'Hash' => $hash
+                ))->first();
+
+            if (!$rememberLoginHash) {
+                $member = null;
+            } else {
+                // Check for expired token
+                $expiryDate = new \DateTime($rememberLoginHash->ExpiryDate);
+                $now = DBDatetime::now();
+                $now = new \DateTime($now->Rfc2822());
+                if ($now > $expiryDate) {
+                    $member = null;
+                }
+            }
+        }
+
+        if ($member) {
+            if ($this->cascadeLogInTo) {
+                // @todo look at how to block "regular login" triggers from happening here
+                // @todo deal with the fact that the Session::current_session() isn't correct here :-/
+                $this->cascadeLogInTo->logIn($member, false, $request);
+            }
+
+            // @todo Consider whether response should be part of logIn() as well
+
+            // Renew the token
+            if ($rememberLoginHash) {
```

Comment:
Considering the `clearCookis()` method, would this be better off moved to a protected `setCookies()` kind of method?